### PR TITLE
Switch compiler to LLVM/MLIR formatting style

### PR DESCRIPTION
**Pull request by kuhar**
The high-level goal is to better align with the LLVM/MLIR coding style and allow for easier prototyping and code movement between the two. The `runtime/` and `tools/` directories still use the `Google` style, as the LLVM config is not always preferred for C code, and we do not expect much sharing/movement of this code with LLVM.

This adds a new `.clang-format` file local to the `compiler/` directory with the same formatting settings as those used by MLIR. The only divergence is not having special case for include ordering that places llvm includes last. It also reformats this directory using the new formatting config.

I used the following command to reformat the code:
```shell
  cd compiler
  fd -e h -e cpp -e cc -e c | xargs clang-format -i
```

I plan to add this commit to `.git-blame-ignore-revs` in a follow-up PR, once the final commit SHA is known.

Issue: https://github.com/openxla/iree/issues/12866

### CODE REVIEW

File: compiler/src/iree/compiler/API/Internal/Embed.cpp
Comment (GMNGeoffrey) [inline compiler/src/iree/compiler/API/Internal/Embed.cpp]: Ew does LLVM style hardcode that LLVM includes always go last?

**Review by GMNGeoffrey — COMMENTED** [review]
Ok I kind of hate that LLVM doesn't indent some nested structures, but whatever consistency with upstream is more useful here.

I don't have a strong opinion on whether we do this in one big PR or leave it to be incremental. On incremental, I think the class keyword indenting is going to be weird if it's inconsistent. Not sure if clang-format will move both `public` and `private` if you touch either or what... Include sorting also seems like it will obscure the actual changes in PRs. Other than that, things look pretty local. Not sure when that very important number of spaces before a namespace comment thing will actually get modified. Seems like there will be a long tail, but also, I don't think we care about that.

File: compiler/src/iree/compiler/API/Internal/Embed.cpp
Comment (kuhar) [reply compiler/src/iree/compiler/API/Internal/Embed.cpp]: It's possible to disable this if we wanted to with [`IncludeCategories`](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#includecategories).

An example: https://github.com/GPUOpen-Drivers/llpc/blob/5ba112ea7a0a394ec2c25d394380bfb33af77600/.clang-format#L5

Comment (GMNGeoffrey) [reply compiler/src/iree/compiler/API/Internal/Embed.cpp]: I'm kind of inclined to do that. That doesn't affect code snippets between projects and if something is getting moved it probably needs its includes tweaked anyway. It seems like something that projects commonly modify. Is there a way to actually *view* what a `BasedOnStyle` does?

Comment (kuhar) [reply compiler/src/iree/compiler/API/Internal/Embed.cpp]: What ordering would we prefer? The existing one or something close to that?

> Is there a way to actually view what a BasedOnStyle does?

Yes, `--dump-config` does that. I uploaded both dumps here: https://gist.github.com/kuhar/c8aa91e494b22fc7a60355fdbdf6ad70.

Comment (GMNGeoffrey) [reply compiler/src/iree/compiler/API/Internal/Embed.cpp]: IDK I'm probably not the right one to decide that. The Google one looks nice and generic just putting system headers and the "main" header first. The LLVM one looks pretty project-specific.

Comment (kuhar) [reply compiler/src/iree/compiler/API/Internal/Embed.cpp]: I switched back to the previous include ordering as it also looked odd to me to be making a special case for llvm headers. This does not compromise on the goal of easier code sharing/movement, and reduced the amount of changes overall.

**Comment by kuhar** [comment]
For the context, we discussed the re-formatting strategy on https://groups.google.com/g/iree-discuss/c/E2-HK557QGc.

The discussion converged on doing this in one shot, like this PR originally started. If everyone's onboard, I'd like to submit it over this weekend to minimize disruptions.

(I'm aware GitHub warns about merge conflicts. I will rebase and run formatting just before submitting, this is completely mechanical.)

Can I get review from key code owners? Paging: @stellaraccident @benvanik @MaheshRavishankar @dcaballe  @GMNGeoffrey @antiagainst

File: compiler/src/iree/compiler/API/Internal/Embed.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/API/Internal/Embed.cpp]: I think the rationale is that includes are sorted from more specific to more generic to avoid missing dependencies. For example:
```
#include <system>
#include "llvm.h"
```
This would compile even if "llvm.h" needs `<system>` but it's accidentally not included in "llvm.h". Given that most of the project depends on llvm and not the other way around, I guess it makes sense to move llvm include right before system. I'm just guessing here, though :)

**Comment by dcaballe** [comment]
I can't actually open the code. My browser hangs but LGTM :)

**Review by dcaballe — APPROVED** [review]

File: compiler/src/iree/compiler/API/Internal/Embed.cpp
Comment (ScottTodd) [reply compiler/src/iree/compiler/API/Internal/Embed.cpp]: https://llvm.org/docs/CodingStandards.html#include-style has more info

**Review by stellaraccident — APPROVED** [review]

File: compiler/src/iree/compiler/API/test/api-test-main.c
Comment (ScottTodd) [inline compiler/src/iree/compiler/API/test/api-test-main.c:1]: This C file might have been intentionally using left pointer alignment to match other C API code in the runtime. Probably makes sense to follow LLVM/MLIR style anyways though.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp:109]: Weird formatting / line breaks here. Could use braces around the conditions and/or comments to force newlines in the returned values.

**Review by ScottTodd — APPROVED** [review]
Scanned through ~30% of the files, looks good.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp:109]: I will fix this in a follow-up PR

**Review by jpienaar — APPROVED** [review]
More lines than expected, it looked closer :-)

File: compiler/.clang-format
Comment (ScottTodd) [inline compiler/.clang-format:17]: I had to update my clang-format for this to work, FYI.
```
YAML:17:22: error: unknown key 'CaseSensitive'
    CaseSensitive:   false
                     ^~~~~
Error reading d:\dev\projects\iree\compiler\.clang-format: invalid argument
```

```
D:\dev\projects\iree (main -> upstream)
λ which clang-format
/c/Program Files/LLVM/bin/clang-format

D:\dev\projects\iree (main -> upstream)
λ clang-format --version
clang-format version 11.0.0
```

(updating to 16.0.0 from https://github.com/llvm/llvm-project/releases/tag/llvmorg-16.0.0 worked)

**Pull request merged**
